In [1]:
import numpy as np

# Section 1

## Question 1

What's the version of NumPy that you installed?

In [2]:
np.__version__

'1.25.2'

**Answer** = 1.25.2

# Section 2

In [3]:
import numpy as np
import pandas as pd

## Question 2

How many records are in the dataset?

In [4]:
data = pd.read_csv('data/car-price-data.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

In [5]:
len(data)

11914

**Answer** = 11914

## Question 3

Who are the most frequent car manufacturers (top-3) according to the dataset?

In [6]:
data.Make.value_counts()[:10]

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Toyota         746
Dodge          626
Nissan         558
GMC            515
Honda          449
Mazda          423
Cadillac       397
Name: count, dtype: int64

**Answer**  =Chevrolet, Ford, Volkswagen

## Question 4

What's the number of unique Audi car models in the dataset?

In [7]:
data[data['Make']=='Audi'].Model.nunique()

34

**Answer** = 34

## Question 5

How many columns in the dataset have missing values?

In [8]:
(
    data.isnull()  # retuns True for each item in DataFrame that is null, False otherwise
    .sum()  # sums up all True values in a column
    .gt(0)  # returns True for all columns with nulls (Greater Than 0)
    .sum()  
)

5

**Answer** = 5

## Question 6

### 6.1
Find the median value of "Engine Cylinders" column in the dataset.

In [9]:
(
    data["Engine Cylinders"].median()
)

6.0

### 6.2
Next, calculate the most frequent value of the same "Engine Cylinders".

In [10]:
(
    data["Engine Cylinders"].value_counts()  # counts number of occurences for each distinct value in the column
    .head()
)

Engine Cylinders
4.0     4752
6.0     4489
8.0     2031
12.0     230
5.0      225
Name: count, dtype: int64

In [11]:
mode_engine_cylinders = data["Engine Cylinders"].mode()[0]
mode_engine_cylinders

4.0

### 6.3
Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.

In [12]:
data["Engine Cylinders"] = (
    data["Engine Cylinders"].fillna(mode_engine_cylinders)
)

### 6.4
Now, calculate the median value of "Engine Cylinders" once again.

In [13]:
(
    data["Engine Cylinders"].median()
)

6.0

Has it changed?

**Answer**: No

# Section 3

In [14]:
from numpy.linalg import inv

## Question 7

### 7.1
Select all the "Lotus" cars from the dataset.

In [15]:
lotus_cars = (
    data[data.Make == 'Lotus']
)

lotus_cars.head()

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   
3915  Exotic,High-Performance      Compact   Convertible           27   
3916  Exotic,High-Performance      Compact   Convertible           26   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250  
3915        21         613  47250  
3916        20         613  54990

### 7.2
Select only columns "Engine HP", "Engine Cylinders".

In [16]:
columns = ["Engine HP", "Engine Cylinders"]
lotus_cars = lotus_cars[columns]
lotus_cars.head()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0

### 7.3 
Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).

In [17]:
lotus_cars.drop_duplicates()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

### 7.4
Get the underlying NumPy array. Let's call it X.

In [18]:
X = lotus_cars.drop_duplicates().values
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

### 7.5 
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [19]:
XTX = X.T.dot(X)  # dot product is the matrix multiplication
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

### 7.6
Compute the inverse of XTX.

In [20]:
XTX_inv = inv(XTX)
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

### 7.7 
Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].

In [21]:
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]
y

[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

### 7.8 
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

In [22]:
print("Shapes of tensors:")
print(f"XTX_inv = {XTX_inv.shape}")
print(f"X.T = {X.T.shape}")
print(f"y = {np.array(y).shape}")

Shapes of tensors:
XTX_inv = (2, 2)
X.T = (2, 9)
y = (9,)


In [23]:
w = (
    XTX_inv.dot(X.T)  # returns a matrix of size 2, 9
    .dot(y)  # returns a vector of length 2
)
w

array([  4.59494481, -63.56432501])

### 7.9
What's the value of the first element of w?

In [24]:
w[0]

4.594944810094551

**Answer**: 4.5949